pkg/controllers/subnetbinding/controller.go

```diff
@@ -0,0 +1,333 @@
+	msgGetSubnetCR              = "Unable to get Subnet CR %s"
+	msgGetSubnetSetCR           = "Unable to get SubnetSet CR %s"
+	msgGetNSXSubnetsBySubnet    = "Subnet CR %s is not realized on NSX"
+	msgGetNSXSubnetsBySubnetSet = "SubnetSet CR %s is not realized on NSX"
+	msgChildWorkAsParent        = "Subnet CR %s is working as target by %s"
+	msgParentWorkAsChild        = "Target Subnet CR %s is attached by %s"
+	msgRealizeSubnetBinding     = "Failed to realize SubnetConnectionBindingMap %s on NSX"
+)
+
+// Reconciler reconciles a SubnetConnectionBindingMap object
+type Reconciler struct {
+	Client               client.Client
+	Scheme               *runtime.Scheme
+	SubnetService        *subnet.SubnetService
+	SubnetBindingService *subnetbinding.BindingService
+	StatusUpdater        common.StatusUpdater
+}
+
+func StartSubnetBindingController(mgr ctrl.Manager, subnetService *subnet.SubnetService, subnetBindingService *subnetbinding.BindingService) {
+	reconciler := newReconciler(mgr, subnetService, subnetBindingService)
+	// Start the controller
+	if err := reconciler.setupWithManager(mgr); err != nil {
+		log.Error(err, "Failed to create controller", "controller", "SubnetConnectionBindingMap")
+		os.Exit(1)
+	}
+	// Start garbage collector in a separate goroutine
+	go common.GenericGarbageCollector(make(chan bool), servicecommon.GCInterval, reconciler.CollectGarbage)
+}
+
+func newReconciler(mgr ctrl.Manager, subnetService *subnet.SubnetService, subnetBindingService *subnetbinding.BindingService) *Reconciler {
+	recorder := mgr.GetEventRecorderFor("subnetconnectionbindingmap-controller")
+	// Create the SubnetConnectionBindingMap Reconciler with the necessary services and configuration
+	return &Reconciler{
+		Client:               mgr.GetClient(),
+		Scheme:               mgr.GetScheme(),
+		SubnetService:        subnetService,
+		SubnetBindingService: subnetBindingService,
+		StatusUpdater:        common.NewStatusUpdater(mgr.GetClient(), subnetBindingService.NSXConfig, recorder, MetricResTypeSubnetConnectionBindingMap, "SubnetConnectionBindingMap", "SubnetConnectionBindingMap"),
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	startTime := time.Now()
+	defer func() {
+		log.Info("Finished reconciling SubnetConnectionBindingMap", "SubnetConnectionBindingMap", req.NamespacedName, "duration(ms)", time.Since(startTime).Milliseconds())
+	}()
+
+	r.StatusUpdater.IncreaseSyncTotal()
+
+	bindingMapCR := &v1alpha1.SubnetConnectionBindingMap{}
+	if err := r.Client.Get(ctx, req.NamespacedName, bindingMapCR); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Try to delete stale NSX SubnetConnectionBindingMaps if exists
+			if err := r.SubnetBindingService.DeleteSubnetConnectionBindingMapsByCRName(req.Name, req.Namespace); err != nil {
+				log.Error(err, "Failed to delete NSX SubnetConnectionBindingMap", "SubnetConnectionBindingMap", req.NamespacedName)
+				return ResultRequeue, nil
```

Comment:
Returns `nil` in Reconcile function even if an error happened because we want to use the ResultRequeue to control the delay to re-queue the request rather than the framework's default value. Otherwise, it may print such logs.
```
2024-11-19 08:16:55.803 INFO    controller/controller.go:314    Warning: Reconciler returned both a non-zero result and a non-nil error. The result will always be ignored if the error is non-nil and the non-nil error causes reqeueuing with exponential backoff. For more details, see: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler        {"controller": "subnetconnectionbindingmap", "controllerGroup": "crd.nsx.vmware.com", "controllerKind": "SubnetConnectionBindingMap", "SubnetConnectionBindingMap": {"name":"binding1","namespace":"svc-tkg-domain-c10"}, "namespace": "svc-tkg-domain-c10", "name": "binding1", "reconcileID": "b504bdcb-9a6b-4009-bc02-f00db6a9dfd1"}
2024-11-19 08:16:55.803 ERROR   controller/controller.go:316    Reconciler error        {"controller": "subnetconnectionbindingmap", "controllerGroup": "crd.nsx.vmware.com", "controllerKind": "SubnetConnectionBindingMap", "SubnetConnectionBindingMap": {"name":"binding1","namespace":"svc-tkg-domain-c10"}, "namespace": "svc-tkg-domain-c10", "name": "binding1", "reconcileID": "b504bdcb-9a6b-4009-bc02-f00db6a9dfd1", "error": "no existing NSX VpcSubnet created by SubnetSet CR 'svc-tkg-domain-c10/pod-default'"}
```